markdown/docs/addon-author-guide.md

```diff
@@ -0,0 +1,321 @@
+environment.
+
+Remember that when updating your addons for FastBoot compatibility,
+your code now must run in both a browser *and* Node.js. Most
+importantly, that means that you cannot just assume that there will
+always be DOM APIs available.
+
+## Don't Break the Boot
+
+The first step towards FastBoot support is ensuring that your addon
+doesn't cause the app to crash when run in Node. You can verify this by
+adding your addon to a new Ember app, adding the FastBoot addon, and
+running the FastBoot server with `ember fastboot`.
+
+If that process results in an error and the server crashing, you know
+that your addon is trying to do something that may work in the browser
+but not in Node.js.
+
+### Common Causes
+
+The most common cause of failure on boot is initialization code that
+relies on the following global objects that aren't available in Node.js:
+
+* `document`
+* `navigator`
+* `window`
+
+For example, many addons may detect features by creating an element via
+`document.createElement()` and checking whether certain properties on it
+exist. Or an addon may try to detect the height of the viewport via
+`window.innerHeight`, but because there is no viewport conceptually in
+Node, this value is not set.
+
+### Tracking It Down
+
+FastBoot works by turning an Ember application and all of its
+dependencies (including your addon) into a single bundle of JavaScript.
+Similar to how you ship an `app.js` and `vendor.js` file to the browser
+when you deploy an Ember app, the FastBoot server also loads these two
+files.
+
+What this means in practice is that errors will include stack traces
+with very large line numbers. Your addon's code will be intermingled
+with other addons, and Ember itself, in the `vendor.js` file.
+
+To figure out what is causing the exception to be thrown, look at the
+stack trace shown when running the app. Note the line number, then open
+`dist/fastboot/vendor.js` in your text editor. Scroll down to the
+provided line number (or, better, use your editor's "Go to Line"
+functionality) to pinpoint exactly which line is causing the issue.
+
+![Example error message and stack trace from an incompatible addon](/images/addon-author-guide/stack-trace-example.png)
+
+### How to Fix It
+
+There are two common approaches to fixing code that accesses browser
+APIs.
+
+First, you can check for the existence of the API you need before
+accessing it. For example, imagine we have an addon that checks the user
+agent of the browser to detect the platform. That code looks something
+like this:
+
+```js
+var ua = window.navigator.userAgent || '';
+```
+
+This line will throw an exception, because `window.navigator` doesn't
+exist in Node. Instead, we can add an extra guard to verify it exists,
+and default to an empty string if not:
+
+```js
+var ua = window && window.navigator ? window.navigator.userAgent : '';
+```
+
+The second option is to move any code that relies on browser-only APIs to
```

Comment:
are these hooks documented? if so, linking here would be helpful, if not, we should document them and link them. lemme know if I can help :)